### PR TITLE
Validate frame attributes when setting them, not when getting them

### LINKED ISSUE #13481: FrameAttribute validation should happen only once
### Description

One more issue from profiling coordinate transformation code:

Everytime on `Frame.__getattr__` the descriptor calls the potentially quite expensive `convert_input` function. This functions is for some attributes even expensive if no transformation is done.  

One way to fix this would be to store the `converted` attribute in the `FrameAttribute` to fully avoid any checks for the case the attribute has already been converted.

### PR DESCRIPTION
**Pull request by maxnoe**
### Description

This pull request adds methods `validate` and `broadcast` to `FrameAttribute` and removes all expensive checks from `__get__`.

`BaseCoordinateFrame` and `SkyCoord` now call `validate` / `convert_input` and `broadcast` where appropriate, so that these operations only happen once, not on each get.

Timings:

This branch:
```
296 ns ± 9.44 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
location
305 ns ± 4.76 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
temperature
319 ns ± 6.95 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

main:
```
❯ ipython time_frame_attr.py
obstime
1.61 µs ± 27.3 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
location
675 ns ± 11.6 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
temperature
1.81 µs ± 14.2 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

Script:

```
from astropy.coordinates import EarthLocation, AltAz
from astropy.time import Time

loc = EarthLocation.of_site("Roque de los Muchachos")
time = Time.now()
frame = AltAz(location=loc, obstime=time)

for attr in ('obstime', 'location', 'temperature'):
    print(attr)
    get_ipython().run_line_magic('timeit', f'frame.{attr}')
```

Fixes #<Issue Number>

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: astropy/coordinates/attributes.py [hunk 1]
```diff
@@ -92,39 +92,51 @@ def convert_input(self, value):
         """
         return value, False
 
+
```
Comment (pllim) [inline astropy/coordinates/attributes.py:95]:
```suggestion
```

**Review by pllim — COMMENTED** [review]
Need a change log.

Since you added new methods and appear public, please add docstring.

And feel free to ping specific coordinates maintainer(s) to review. Thanks!

**Comment by maxnoe** [comment]
@mhvk Want to take a look? I am also not sure if this is something that can go into 5.2 as @pllim set for now, this might need to wait for 6.0?

**Comment by pllim** [comment]
Re: 5.2 or 6.0 -- Up to you. If you want it in 6.0, this can never be merged before 5.2.x branching, so safer to turn it back into draft.

Re: breaking sunpy -- Maybe @Cadair can ping his Sunpy peeps to have a look.

**Comment by Cadair** [comment]
This is definitely going to need some sunpy changes, what worries me from that perspective is it looks pretty hard to support both the old and new code at the same time, which is what sunpy would need to do.

I would need to experiment to see how we could adapt.

@ayshih ping.

**Comment by mhvk** [comment]
Note that we could avoid repeated validation also by memoizing the outcome. Though I find the logic that is being introduced here, or doing validation on setting, and broadcasting later, a bit more pleasant conceptually. Anyway, it would be good to know in a bit more detail what would go wrong with this in sunpy; we should avoid adding a lot of work if we can solve the performance issue differently.

**Comment by Cadair** [comment]
Without looking in detail, it's just that we have our Attrs which do validation on get. We would need to adapt and support old and new at the same time. That's as far as I have looked.

**Comment by ayshih** [comment]
I haven't actually run against this PR, but I suspect we don't need to make any changes in SunPy.  It'll just end up that one of our frame attributes with an overridden `__get__()` will still sometimes validate on getting in addition to on setting (because it can't fully convert the value without accessing a different frame attribute), and I don't see a way around that.

**Comment by ayshih** [comment]
Now tested: I don't encounter any conflicts with SunPy's existing code.

The changes in this PR are cleaner logic, but I'll point out that the performance improvements are very small.  That's because, in the existing implementation, once the value has been converted and broadcasted on the first call to `__get__()`, it shouldn't be converted or broadcasted again (assuming reasonable definitions of `convert_input()` that bail out quickly).

**Comment by maxnoe** [comment]
>  That's because, in the existing implementation, once the value has been converted and broadcasted on the first call to __get__(), it shouldn't be converted or broadcasted again (assuming reasonable definitions of convert_input() that bail out quickly).

That's obviously not the case for the ones I tested, see the benchmark code

**Comment by mhvk** [comment]
For the secondary attribute, I'm not sure we have a test...  Right now, it seems we only use it in `FK4`, for
```
    obstime = TimeAttribute(default=None, secondary_attribute='equinox')
```
validation for `obstime` and `equinox` is the same anyway.

But given that we use it so little, perhaps it is OK to assume validation of the secondary attribute is fine.

File: astropy/coordinates/attributes.py [hunk 2]
```diff
@@ -92,39 +92,64 @@ def convert_input(self, value):
...
+        name = f"_{self.name}"
+        out = getattr(instance, name)
+        out_shape = getattr(out, 'shape', ())
+
+        instance_shape = getattr(instance, 'shape', None)  # None if instance (frame) has no data!
+        converted = False
+        if instance_shape is not None and out_shape != () and out_shape != instance_shape:
+            # If the shapes do not match, try broadcasting.
+            try:
+                if isinstance(out, ShapedLikeNDArray):
```
Comment (mhvk) [inline astropy/coordinates/attributes.py:127]: Bit of an aside, but we can remove this `if/else` -- `ShapedLineNDArray` will work with `np.broadcast_to`.

File: astropy/coordinates/attributes.py [hunk 3]
```diff
@@ -92,39 +92,64 @@ def convert_input(self, value):
...
+                else:
+                    out = np.broadcast_to(out, instance_shape, subok=True)
+            except ValueError:
+                # raise more informative exception.
+                raise ValueError(
+                    "attribute {} should be scalar or have shape {}, "
+                    "but is has shape {} and could not be broadcast."
+                    .format(self.name, instance_shape, out.shape))
+
+            converted = True
```
Comment (mhvk) [inline astropy/coordinates/attributes.py:139]: Can just do `setattr(instance, name, out)` here - it is the only place where `converted` gets set (unlike before, when `convert_input` could set it).

**Review by mhvk — COMMENTED** [review]
Overall, I feel this is a more logical approach - the fact that one doesn't need comments to explain what is happening is a very good sign! Just a few small comments.

Also a comment for future-proofing - I suggest we use private methods `_validate` and `_broadcast`, mostly because I think they are implementation details that may well change. In particular, seeing this I feel the idea solution would be to use `__set__` for setting and validation (making it readonly after initial initialization, like what @nstarman and I ended up with for attributes of cosmological models). Now I definitely don't think we should wait with this PR for that, but it seems good to ensure that it is clear for downstream packages that `convert_input` is the only part one should adjust.

**Comment by nstarman** [comment]
Oops @maxnoe, do you want to re-open this PR?

**Comment by maxnoe** [comment]
Yes, I still think this is a good idea, however I cannot promise when I can get back to it

**Pull request closed without merging**